include/validator/validator_component.h

```diff
@@ -148,9 +165,23 @@ struct InstanceExprVisitor {
     if (std::holds_alternative<CoreSort>(ArgSort)) {
       CoreSort ArgCoreSort = std::get<CoreSort>(ArgSort);
 
-      if (std::holds_alternative<DescTypeIndex>(ImportType) &&
-          ArgCoreSort == CoreSort::Type) {
-        return true;
+      if (std::holds_alternative<DescTypeIndex>(ImportType)) {
+        const auto &Desc = std::get<DescTypeIndex>(ImportType);
+        IndexKind Kind = Desc.getKind();
+
+        if ((Kind == IndexKind::CoreType && ArgCoreSort == CoreSort::Type) ||
+            (Kind == IndexKind::FuncType && ArgCoreSort == CoreSort::Func) ||
+            (Kind == IndexKind::ComponentType &&
+             ArgCoreSort == CoreSort::Module) ||
+            (Kind == IndexKind::InstanceType &&
+             ArgCoreSort == CoreSort::Instance)) {
+          return true;
+        }
+        spdlog::error("[Core Sort] Type mismatch: Expected '{}' but got '{}'",
+                      WasmEdge::Validator::toString(Kind),
+                      WasmEdge::Validator::toString(ArgCoreSort));
+
+        return false;
       }
 
       if (std::holds_alternative<ValueType>(ImportType)) {
```

Comment:
each `holds_alternative` should be `else if` even you use early return, since for future that will help refactoring (if needed) easier, and everyone can see these branches are the same level.